### PR TITLE
Adding robust noise model for measurement noise

### PR DESCRIPTION
**Pull request by ayushbaid**
The input data for bundle adjustment occasionally contains outliers in tracks or 3d points. A robust noise model will help us tackle these outliers.

In this PR, I add a boolean option to switch to huber function based noise for 2d measurement factors. 

Right now, this PR has merged in changes from #65. I'll add the changes from that PR too.

### CODE REVIEW

**Comment by johnwlambert** [comment]
Looks like we need to pull master into here

**Comment by ayushbaid** [comment]
Metrics:

| Metric | Master | Individual intrinsics + Non Robust Noise | Shared intrinsics + Non Robust Noise |  Individual intrinsics + Robust Noise | Shared intrinsics + Robust Noise |
| ----- | ----- | ---- | ---- | ---- | ----- | 
| BA Input num tracks | 9169 | 9159 | 9178 | 9150 | 9165 | 
| Initial Error | 1.26e6 | 1.03e6 | 5.21e6 | 1.46e5 | 2.08e5 |
| Final Error | 7.88e4 | 6.04e4 | 7.03e4 | 2.09e3 | 2.00e3 | 
| Num tracks post filtering | 9116 | 9108 | 9126 | 9091 | 9109 |
| Mean track length | 5.243 | 5.246 | 5.238 | 5.249 | 5.240 | 
| Median track length | 4.000 | 4.000 | 4.000 | 4.000 | 4.000 |
| Avg scene reproj error | 0.197 | 0.207 | 0.212 | 0.180 | 0.186 |
| Median scene reproj error | 0.158 | 0.165 | 0.172 | 0.140 | 0.148 |
| Max scene reproj error | 2.986 | 2.970 | 2.988 | 2.988 | 2.958 |

**Comment by johnwlambert** [comment]
@ayushbaid please revisit

**Comment by ayushbaid** [comment]
# Skydio Crane 32 dataset

| Metric name                            |       Baseline | New      |
|--------------------------------------|---------------|----------|
| number_cameras                         |      32        | 32       |
| number_tracks_unfiltered               |    5225        | 5099     |
| mean_3d_track_lengths_unfiltered       |       2.65722  | 2.6564   |
| mean_reprojection_errors_unfiltered_px |       7.33123  | 16.984   |
| mean_rotation_error_angle_deg          |       9.02262  | 7.00743  |
| mean_translation_error_distance        |       0.526999 | 0.439402 |
| mean_translation_angle_error_deg       |       8.81846  | 7.32207  |
| number_tracks_filtered                 |    2236        | 1647     |
| mean_3d_track_lengths_filtered         |       2.10063  | 2.07347  |
| mean_reprojection_errors_filtered_px   |       1.31008  | 1.29176  |

**Baseline**
baseline_ba_error

**New**
new_ba_error

**Comment by ayushbaid** [comment]
@akshay-krishnan  The histogram for 3D track lengths seem off

File: gtsfm/bundle/bundle_adjustment.py [hunk 1]
```diff
@@ -62,7 +72,13 @@ def run(
             return initial_data
 
         # noise model for measurements -- one pixel in u and v
-        measurement_noise = gtsam.noiseModel.Isotropic.Sigma(IMG_MEASUREMENT_DIM, 1.0)
+        if self._robust_measurement_noise:
+            measurement_noise = gtsam.noiseModel.Robust(
+                gtsam.noiseModel.mEstimator.Huber(1.35),
+                gtsam.noiseModel.Isotropic.Sigma(IMG_MEASUREMENT_DIM, 1.0),
+            )
```
Comment (ayushbaid) [inline gtsfm/bundle/bundle_adjustment.py:79]: is this the correct way to do it? I am using the same object of noise model in all factors. Should I create a copy every time? @ProfFan

**Comment by johnwlambert** [comment]
Thanks Ayush. Looks like the pose errors after BA go down, which seems like an improvement. Maybe we can discuss more in the meeting?

**Comment by johnwlambert** [comment]
I wonder if the improvement is reproducible after a few runs, or just random fluctuation?

File: gtsfm/configs/scene_optimizer_unit_test_config.yaml [hunk 2]
```diff
@@ -25,7 +25,7 @@ SceneOptimizer:
       estimation_threshold_px: 0.5 # for H/E/F estimators
       min_allowed_inlier_ratio_est_model: 0.1
 
-  multiview_optimizer:
+  multiview_optimizer:True
```
Comment (johnwlambert) [inline gtsfm/configs/scene_optimizer_unit_test_config.yaml:28]: why do we put True here in the config?

File: gtsfm/bundle/bundle_adjustment.py (hunk 1 above)
Comment (ProfFan) [reply gtsfm/bundle/bundle_adjustment.py:79]: I don't think you need to do that

File: gtsfm/bundle/bundle_adjustment.py [hunk 3]
```diff
@@ -62,7 +72,13 @@ def run(
             return initial_data
 
         # noise model for measurements -- one pixel in u and v
-        measurement_noise = gtsam.noiseModel.Isotropic.Sigma(IMG_MEASUREMENT_DIM, 1.0)
+        if self._robust_measurement_noise:
+            measurement_noise = gtsam.noiseModel.Robust(
+                gtsam.noiseModel.mEstimator.Huber(1.35),
```
Comment (johnwlambert) [inline gtsfm/bundle/bundle_adjustment.py:77]: can we try with this and the default?

**Comment by johnwlambert** [comment]
We should check the median for reproj error (outliers...

**Comment by johnwlambert** [comment]
@ayushbaid I saw that ` tests/data/door_initialization.bal` was checked into this PR. What do we need it for?

**Comment by johnwlambert** [comment]
Can we pull in master one more time? Looks like `
```
FAILED tests/test_scene_optimizer.py::TestSceneOptimizer::test_create_computation_graph
```
is failing

File: gtsfm/bundle/bundle_adjustment.py (hunk 3 above)
Comment (ayushbaid) [reply gtsfm/bundle/bundle_adjustment.py:77]: The hyperparameter in Huber is defined in terms of number of sigma. Everywhere in GTSAM, the value of 1.345 is used. I found a reference for it: https://myweb.uiowa.edu/pbreheny/uk/764/notes/12-1.pdf. There is no default value. If we want to experiment, I think the hyperparameter to change will be the sigma in the noise model. 1 seems too small

https://github.com/borglab/gtsam/blob/43e8f1e5aeaf11890262722c1e5e04a11dbf9d75/gtsam/linear/NoiseModel.h#L651.

**Comment by ayushbaid** [comment]
> @ayushbaid I saw that ` tests/data/door_initialization.bal` was checked into this PR. What do we need it for?

removed it.

File: gtsfm/configs/scene_optimizer_unit_test_config.yaml (hunk 2 above)
Comment (ayushbaid) [reply gtsfm/configs/scene_optimizer_unit_test_config.yaml]: removed it.

**Comment by ayushbaid** [comment]
> We should check the median for reproj error (outliers...

From the plots, the median reprojection error goes down.

**Review by johnwlambert — APPROVED** [review]
LGTM. Thanks Ayush!

**Pull request merged**